[14.0] shopfloor_manual_product_transfer: new warning about reserved qty

sebalix opened the pull request:
... and display processed quantity with the `packaging-qty-picker` in readonly mode + some minor improvements/fixes.

![image](https://user-images.githubusercontent.com/5315285/201111335-cd6d2062-3899-4916-a4c6-ee1661e8f45d.png)

Ref: rau-67

sebalix committed:
sf_m_p_transfer: move check in a method

Move code in a method '_check_quantity_in_stock'.

sebalix commented on shopfloor_manual_product_transfer/services/manual_product_transfer.py:
```diff
@@ -511,9 +521,17 @@ def _unreserve_other_lines(self, location, move_lines):
             [
                 ("location_id", "=", location.id),
                 ("state", "in", ("assigned", "partially_available")),
+                # FIXME: maybe the answer to the TODO below: should we not
+                # add 'product_id' in the domain here? We could block the entire
+                # process here (see check on 'extra_move_lines' below) because
+                # another move line for another product is found in the given
+                # location... is it what we want?
             ]
         )
         extra_move_lines = location_move_lines - move_lines
+        # TODO: question, why do we care about lines that could be related to
+        # other products here? Should it prevent us to unreserve the moves of
+        # the product we're interested in?
```
ping @jbaudoux I would like your insight on this, I think `location_move_lines` could contain unrelated lines (of other products sharing the same location), maybe it doesn't happen IRL with location bins but to me it's a bug, unless I didn't understood what was the purpose of the unreserve process in this scenario.

jbaudoux replied:
Indeed the method should take in argument an option product. For the location content transfer, it is correct to unreserve all others moves as we empty the location. For this scenario where we move only a product, we should limit to the given product.

sebalix replied:
Fixed here:
- https://github.com/OCA/wms/pull/482

sebalix committed:
sf_m_p_transfer: fix error message on scan_product

sebalix committed:
sf_m_p_transfer: display warning on reserved qty

If the option "Allow to process reserved quantities" is not enabled
we should at least display a warning to the operator to not move
the quantity already reserved.

sebalix committed:
sf_m_p_transfer: improve qty done display

Display the done qty with packaging-qty-picker in readonly.

TDu commented on shopfloor_manual_product_transfer/services/manual_product_transfer.py:
```diff
@@ -304,6 +324,17 @@ def _check_quantity(self, location, product, lot, quantity):
+        current_qty = self._get_product_qty(location, product, lot, free=True)
+        initial_qty = self._get_initial_qty(location, product, lot)
+        return (
+            # FIXME use float_compare
```
Probably a good time to fix it.

sebalix replied:
Fixed

simahawk commented on shopfloor_manual_product_transfer_mobile/static/wms/src/scenario/manual_product_transfer.js:
```diff
@@ -66,9 +76,11 @@ const ManualProductTransfer = {
-                        <span class="label">Quantity = {{ quantity() }}</span>
-                    </v-card-title>
+                    <packaging-qty-picker
+                        :key="make_state_component_key(['packaging-qty-picker', 1232])"
```
1232?? this should be a uniq value. Use the id of the record if you have one

sebalix replied:
Fixed, everywhere + tested locally.

simahawk commented on shopfloor_manual_product_transfer_mobile/static/wms/src/scenario/manual_product_transfer.js:
```diff
@@ -130,6 +142,16 @@ const ManualProductTransfer = {
             return {};
         },
+        warning: function () {
+            const data = this.state.data;
+            if (_.isEmpty(data)) {
+                return "";
+            }
+            if (!_.isEmpty(data.warning)) {
+                return data.warning;
+            }
+            return "";
```
```suggestion
            return _.result(this.state, "data.warning", "")
```

simahawk replied:
BTW IMO this warning should be defined only frontend side and enabled w/ a flag instead.

sebalix replied:
I think about it, and also think about the translations which are already integrated in the backend (not there are no translations in Shopfloor but it's easier on the backend).

simahawk commented on shopfloor_manual_product_transfer/actions/message.py:
```diff
@@ -25,3 +25,12 @@ def transfer_canceled_success(self, picking):
+        return {
+            "message_type": "warning",
+            "body": _(
+                f"{qty_assigned} {product.uom_id.name} are reserved in this "
```
Are you sure `f` strings work fine w/ translations?

sebalix replied:
Not at all :D fixed

simahawk commented:
And please shorten the commits that are too long ;)

simahawk commented:
/ocabot merge minor
